<!DOCTYPE html>
<html lang='en'>
<head>
<title>idmapd-dnotify-to-inotify.patch\nfs-utils\main - aports - Main aports tree
</title>
<meta name='generator' content='cgit v1.2.1'/>
<meta name='robots' content='index, nofollow'/>
<link rel='stylesheet' type='text/css' href='/cgit.css'/>
<link rel='shortcut icon' href='//alpinelinux.org/alpine-logo.ico'/>
<link rel='alternate' title='Atom feed' href='https://git.alpinelinux.org/aports/atom/main/nfs-utils/idmapd-dnotify-to-inotify.patch?h=master' type='application/atom+xml'/>
<link rel='vcs-git' href='git://git.alpinelinux.org/aports' title='aports Git repository'/>
<link rel='vcs-git' href='https://git.alpinelinux.org/aports' title='aports Git repository'/>
</head>
<body>
<div id='cgit'><table id='header'>
<tr>
<td class='logo' rowspan='2'><a href='/'><img src='//wiki.alpinelinux.org/images/alogo.png' alt='cgit logo'/></a></td>
<td class='main'><a href='/'>index</a> : <a title='aports' href='/aports/'>aports</a></td><td class='form'><form method='get'>
<select name='h' onchange='this.form.submit();'>
<option value='1.10-stable'>1.10-stable</option>
<option value='1.9'>1.9</option>
<option value='2.0-stable'>2.0-stable</option>
<option value='2.1-stable'>2.1-stable</option>
<option value='2.2-stable'>2.2-stable</option>
<option value='2.3-stable'>2.3-stable</option>
<option value='2.4-stable'>2.4-stable</option>
<option value='2.5-stable'>2.5-stable</option>
<option value='2.6-stable'>2.6-stable</option>
<option value='2.7-stable'>2.7-stable</option>
<option value='3.0-stable'>3.0-stable</option>
<option value='3.1-stable'>3.1-stable</option>
<option value='3.2-stable'>3.2-stable</option>
<option value='3.3-stable'>3.3-stable</option>
<option value='3.4-stable'>3.4-stable</option>
<option value='3.5-stable'>3.5-stable</option>
<option value='3.6-stable'>3.6-stable</option>
<option value='3.7-stable'>3.7-stable</option>
<option value='3.8-stable'>3.8-stable</option>
<option value='3.9-stable'>3.9-stable</option>
<option value='master' selected='selected'>master</option>
</select> <input type='submit' value='switch'/></form></td></tr>
<tr><td class='sub'>Main aports tree
</td><td class='sub right'>gitolite</td></tr></table>
<table class='tabs'><tr><td>
<a href='/aports/about/'>about</a><a href='/aports/'>summary</a><a href='/aports/refs/'>refs</a><a href='/aports/log/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>log</a><a class='active' href='/aports/tree/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>tree</a><a href='/aports/commit/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>commit</a><a href='/aports/diff/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>diff</a><a href='/aports/stats/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>stats</a></td><td class='form'><form class='right' method='get' action='/aports/log/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>
<select name='qt'>
<option value='grep'>log msg</option>
<option value='author'>author</option>
<option value='committer'>committer</option>
<option value='range'>range</option>
</select>
<input class='txt' type='search' size='10' name='q' value=''/>
<input type='submit' value='search'/>
</form>
</td></tr></table>
<div class='path'>path: <a href='/aports/tree/'>root</a>/<a href='/aports/tree/main'>main</a>/<a href='/aports/tree/main/nfs-utils'>nfs-utils</a>/<a href='/aports/tree/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>idmapd-dnotify-to-inotify.patch</a></div><div class='content'>blob: d1a3011f7f45528d2dced913d68f553068c150c7 (<a href='/aports/plain/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>plain</a>) (<a href='/aports/blame/main/nfs-utils/idmapd-dnotify-to-inotify.patch'>blame</a>)
<table summary='blob content' class='blob'>
<tr><td class='linenumbers'><pre><a id='n1' href='#n1'>1</a>
<a id='n2' href='#n2'>2</a>
<a id='n3' href='#n3'>3</a>
<a id='n4' href='#n4'>4</a>
<a id='n5' href='#n5'>5</a>
<a id='n6' href='#n6'>6</a>
<a id='n7' href='#n7'>7</a>
<a id='n8' href='#n8'>8</a>
<a id='n9' href='#n9'>9</a>
<a id='n10' href='#n10'>10</a>
<a id='n11' href='#n11'>11</a>
<a id='n12' href='#n12'>12</a>
<a id='n13' href='#n13'>13</a>
<a id='n14' href='#n14'>14</a>
<a id='n15' href='#n15'>15</a>
<a id='n16' href='#n16'>16</a>
<a id='n17' href='#n17'>17</a>
<a id='n18' href='#n18'>18</a>
<a id='n19' href='#n19'>19</a>
<a id='n20' href='#n20'>20</a>
<a id='n21' href='#n21'>21</a>
<a id='n22' href='#n22'>22</a>
<a id='n23' href='#n23'>23</a>
<a id='n24' href='#n24'>24</a>
<a id='n25' href='#n25'>25</a>
<a id='n26' href='#n26'>26</a>
<a id='n27' href='#n27'>27</a>
<a id='n28' href='#n28'>28</a>
<a id='n29' href='#n29'>29</a>
<a id='n30' href='#n30'>30</a>
<a id='n31' href='#n31'>31</a>
<a id='n32' href='#n32'>32</a>
<a id='n33' href='#n33'>33</a>
<a id='n34' href='#n34'>34</a>
<a id='n35' href='#n35'>35</a>
<a id='n36' href='#n36'>36</a>
<a id='n37' href='#n37'>37</a>
<a id='n38' href='#n38'>38</a>
<a id='n39' href='#n39'>39</a>
<a id='n40' href='#n40'>40</a>
<a id='n41' href='#n41'>41</a>
<a id='n42' href='#n42'>42</a>
<a id='n43' href='#n43'>43</a>
<a id='n44' href='#n44'>44</a>
<a id='n45' href='#n45'>45</a>
<a id='n46' href='#n46'>46</a>
<a id='n47' href='#n47'>47</a>
<a id='n48' href='#n48'>48</a>
<a id='n49' href='#n49'>49</a>
<a id='n50' href='#n50'>50</a>
<a id='n51' href='#n51'>51</a>
<a id='n52' href='#n52'>52</a>
<a id='n53' href='#n53'>53</a>
<a id='n54' href='#n54'>54</a>
<a id='n55' href='#n55'>55</a>
<a id='n56' href='#n56'>56</a>
<a id='n57' href='#n57'>57</a>
<a id='n58' href='#n58'>58</a>
<a id='n59' href='#n59'>59</a>
<a id='n60' href='#n60'>60</a>
<a id='n61' href='#n61'>61</a>
<a id='n62' href='#n62'>62</a>
<a id='n63' href='#n63'>63</a>
<a id='n64' href='#n64'>64</a>
<a id='n65' href='#n65'>65</a>
<a id='n66' href='#n66'>66</a>
<a id='n67' href='#n67'>67</a>
<a id='n68' href='#n68'>68</a>
<a id='n69' href='#n69'>69</a>
<a id='n70' href='#n70'>70</a>
<a id='n71' href='#n71'>71</a>
<a id='n72' href='#n72'>72</a>
<a id='n73' href='#n73'>73</a>
<a id='n74' href='#n74'>74</a>
<a id='n75' href='#n75'>75</a>
<a id='n76' href='#n76'>76</a>
<a id='n77' href='#n77'>77</a>
<a id='n78' href='#n78'>78</a>
<a id='n79' href='#n79'>79</a>
<a id='n80' href='#n80'>80</a>
<a id='n81' href='#n81'>81</a>
<a id='n82' href='#n82'>82</a>
<a id='n83' href='#n83'>83</a>
<a id='n84' href='#n84'>84</a>
<a id='n85' href='#n85'>85</a>
<a id='n86' href='#n86'>86</a>
<a id='n87' href='#n87'>87</a>
<a id='n88' href='#n88'>88</a>
<a id='n89' href='#n89'>89</a>
<a id='n90' href='#n90'>90</a>
<a id='n91' href='#n91'>91</a>
<a id='n92' href='#n92'>92</a>
<a id='n93' href='#n93'>93</a>
<a id='n94' href='#n94'>94</a>
<a id='n95' href='#n95'>95</a>
<a id='n96' href='#n96'>96</a>
<a id='n97' href='#n97'>97</a>
<a id='n98' href='#n98'>98</a>
<a id='n99' href='#n99'>99</a>
<a id='n100' href='#n100'>100</a>
<a id='n101' href='#n101'>101</a>
<a id='n102' href='#n102'>102</a>
<a id='n103' href='#n103'>103</a>
<a id='n104' href='#n104'>104</a>
<a id='n105' href='#n105'>105</a>
</pre></td>
<td class='lines'><pre><code>From patchwork Wed Jun 21 21:23:38 2017
Content-Type: text/plain; charset=&quot;utf-8&quot;
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: [V2] idmapd: Use inotify instead of dnotify
From: Alan Swanson &lt;reiver&#64;improbability.net&gt;
X-Patchwork-Id: 9802949
Message-Id: &lt;20170621212338.32460-1-reiver&#64;improbability.net&gt;
To: linux-nfs&#64;vger.kernel.org
Cc: Alan Swanson &lt;reiver&#64;improbability.net&gt;
Date: Wed, 21 Jun 2017 22:23:38 +0100

Remove last use of dnotify in nfs-utils by bringing idmapd upto
date with (required) inotify use by gssd and blkmapd.
<span style="color:#0057ae">---</span>
 utils/idmapd/idmapd.c | 32 +++++++++++++++++---------------
 1 file changed, 17 insertions(+), 15 deletions(-)

Got annoyed by the dnotify kernel requirement. Read the last
discussion &quot;DNOTIFY to INOTIFY migration&quot; posted in 2014 and while
still not officially depreciated, dnotify was officially &quot;replaced&quot;
by inotify in Linux 2.6.13 in 2005. Would be nice to sync this
requirement across the codebase.

V2: Init wd to -1 not 0

diff --git a/utils/idmapd/idmapd.c b/utils/idmapd/idmapd.c
index c12e878..c29dba6 100644
<span style="color:#0057ae">--- a/utils/idmapd/idmapd.c</span>
<span style="color:#000000; font-weight:bold">+++ b/utils/idmapd/idmapd.c</span>
<span style="color:#010181">&#64;&#64; -36,7 +36,7 &#64;&#64;</span>
 
 #include &lt;sys/types.h&gt;
 #include &lt;sys/time.h&gt;
<span style="color:#0057ae">-#include &lt;sys/poll.h&gt;</span>
<span style="color:#000000; font-weight:bold">+#include &lt;sys/inotify.h&gt;</span>
 #include &lt;sys/socket.h&gt;
 #include &lt;sys/stat.h&gt;
 #include &lt;time.h&gt;
&#64;&#64; -205,15 +205,16 &#64;&#64; void usage(char *progname)
 int
 main(int argc, char **argv)
 {
<span style="color:#0057ae">-	int fd = 0, opt, fg = 0, nfsdret = -1;</span>
<span style="color:#000000; font-weight:bold">+	int wd = -1, opt, fg = 0, nfsdret = -1;</span>
 	struct idmap_clientq icq;
<span style="color:#0057ae">-	struct event rootdirev, clntdirev, svrdirev;</span>
<span style="color:#000000; font-weight:bold">+	struct event rootdirev, clntdirev, svrdirev, inotifyev;</span>
 	struct event initialize;
 	struct passwd *pw;
 	struct group *gr;
 	struct stat sb;
 	char *xpipefsdir = NULL;
 	int serverstart = 1, clientstart = 1;
<span style="color:#000000; font-weight:bold">+	int inotify_fd;</span>
 	int ret;
 	char *progname;
 	char *conf_path = NULL;
&#64;&#64; -373,18 +374,15 &#64;&#64; main(int argc, char **argv)
 			}
 		}
 
<span style="color:#0057ae">-		if ((fd = open(pipefsdir, O_RDONLY)) == -1)</span>
<span style="color:#0057ae">-			xlog_err(&quot;main: open(%s): %s&quot;, pipefsdir, strerror(errno));</span>
<span style="color:#0057ae">-</span>
<span style="color:#0057ae">-		if (fcntl(fd, F_SETSIG, SIGUSR1) == -1)</span>
<span style="color:#0057ae">-			xlog_err(&quot;main: fcntl(%s): %s&quot;, pipefsdir, strerror(errno));</span>
<span style="color:#0057ae">-</span>
<span style="color:#0057ae">-		if (fcntl(fd, F_NOTIFY,</span>
<span style="color:#0057ae">-			DN_CREATE | DN_DELETE | DN_MODIFY | DN_MULTISHOT) == -1) {</span>
<span style="color:#0057ae">-			xlog_err(&quot;main: fcntl(%s): %s&quot;, pipefsdir, strerror(errno));</span>
<span style="color:#0057ae">-			if (errno == EINVAL)</span>
<span style="color:#0057ae">-				xlog_err(&quot;main: Possibly no Dnotify support in kernel.&quot;);</span>
<span style="color:#000000; font-weight:bold">+		inotify_fd = inotify_init1(IN_NONBLOCK);</span>
<span style="color:#000000; font-weight:bold">+		if (inotify_fd == -1) {</span>
<span style="color:#000000; font-weight:bold">+			xlog_err(&quot;Unable to initialise inotify_init1: %s\n&quot;, strerror(errno));</span>
<span style="color:#000000; font-weight:bold">+		} else {</span>
<span style="color:#000000; font-weight:bold">+			wd = inotify_add_watch(inotify_fd, pipefsdir, IN_CREATE | IN_DELETE | IN_MODIFY);</span>
<span style="color:#000000; font-weight:bold">+			if (wd &lt; 0)</span>
<span style="color:#000000; font-weight:bold">+				xlog_err(&quot;Unable to inotify_add_watch(%s): %s\n&quot;, pipefsdir, strerror(errno));</span>
 		}
<span style="color:#000000; font-weight:bold">+</span>
 		TAILQ_INIT(&amp;icq);
 
 		/* These events are persistent */
&#64;&#64; -394,6 +392,10 &#64;&#64; main(int argc, char **argv)
 		signal_add(&amp;clntdirev, NULL);
 		signal_set(&amp;svrdirev, SIGHUP, svrreopen, NULL);
 		signal_add(&amp;svrdirev, NULL);
<span style="color:#000000; font-weight:bold">+		if ( wd &gt;= 0) {</span>
<span style="color:#000000; font-weight:bold">+			event_set(&amp;inotifyev, inotify_fd, EV_READ | EV_PERSIST, dirscancb, &amp;icq);</span>
<span style="color:#000000; font-weight:bold">+			event_add(&amp;inotifyev, NULL);</span>
<span style="color:#000000; font-weight:bold">+		}</span>
 
 		/* Fetch current state */
 		/* (Delay till start of event_dispatch to avoid possibly losing
&#64;&#64; -402,7 +404,7 &#64;&#64; main(int argc, char **argv)
 		evtimer_add(&amp;initialize, &amp;now);
 	}
 
<span style="color:#0057ae">-	if (nfsdret != 0 &amp;&amp; fd == 0)</span>
<span style="color:#000000; font-weight:bold">+	if (nfsdret != 0 &amp;&amp; wd &lt; 0)</span>
 		xlog_err(&quot;main: Neither NFS client nor NFSd found&quot;);
 
 	daemon_ready();
</code></pre></td></tr></table>
</div> <!-- class=content -->
<div class='footer'>generated by <a href='https://git.zx2c4.com/cgit/about/'>cgit v1.2.1</a> (<a href='https://git-scm.com/'>git 2.18.0</a>) at 2019-04-01 23:24:45 +0000</div>
</div> <!-- id=cgit -->
</body>
</html>
